vortex-layout/src/layouts/dict/writer/mod.rs

```diff
@@ -0,0 +1,188 @@
+use bytes::Bytes;
+use vortex_array::arcref::ArcRef;
+use vortex_array::compute::slice;
+use vortex_array::vtable::EncodingVTable as _;
+use vortex_array::{Array, ArrayContext, ArrayRef, RkyvMetadata, SerializeMetadata};
+use vortex_btrblocks::BtrBlocksCompressor;
+use vortex_dict::DictEncoding;
+use vortex_dict::builders::{DictConstraints, DictEncoder, dict_encoder};
+use vortex_dtype::{DType, PType};
+use vortex_error::{VortexResult, vortex_bail, vortex_err};
+
+mod repeating;
+mod single;
+
+use crate::layouts::dict::DictLayout;
+use crate::{Layout, LayoutStrategy, LayoutVTableRef, LayoutWriter, LayoutWriterExt};
+
+#[derive(Clone)]
+pub struct DictLayoutOptions {
+    pub repeat: bool,
+    pub constraints: DictConstraints,
+}
+
+impl Default for DictLayoutOptions {
+    fn default() -> Self {
+        Self {
+            repeat: true,
+            constraints: DictConstraints {
+                max_bytes: 1024 * 1024,
+                max_len: u16::MAX as usize,
+            },
+        }
+    }
+}
+
+pub struct DictStrategy {
+    pub options: DictLayoutOptions,
+    pub child: ArcRef<dyn LayoutStrategy>,
+    pub values: ArcRef<dyn LayoutStrategy>,
+}
+
+impl LayoutStrategy for DictStrategy {
+    fn new_writer(&self, ctx: &ArrayContext, dtype: &DType) -> VortexResult<Box<dyn LayoutWriter>> {
+        if !dict_layout_supported(dtype) {
+            return self.child.new_writer(ctx, dtype);
+        }
+        Ok(DelegatingDictLayoutWriter::Uninit(DictWriterArgs {
+            ctx: ctx.clone(),
+            options: self.options.clone(),
+            child_strategy: self.child.clone(),
+            values_strategy: self.values.clone(),
+            dtype: dtype.clone(),
+        })
+        .boxed())
+    }
+}
+
+pub fn dict_layout_supported(dtype: &DType) -> bool {
+    matches!(
+        dtype,
+        DType::Primitive(..) | DType::Utf8(_) | DType::Binary(_)
+    )
+}
+
+struct DictWriterArgs {
+    ctx: ArrayContext,
+    options: DictLayoutOptions,
+    child_strategy: ArcRef<dyn LayoutStrategy>,
+    values_strategy: ArcRef<dyn LayoutStrategy>,
+    dtype: DType,
+}
+
+enum DelegatingDictLayoutWriter {
+    Uninit(DictWriterArgs),
+    Writer(Box<dyn LayoutWriter>),
+}
+
+impl LayoutWriter for DelegatingDictLayoutWriter {
+    fn push_chunk(
+        &mut self,
+        segment_writer: &mut dyn crate::segments::SegmentWriter,
+        chunk: ArrayRef,
+    ) -> VortexResult<()> {
+        match self {
+            Self::Uninit(args) => {
+                let compressed = BtrBlocksCompressor.compress(&chunk)?;
+                *self = if !compressed.is_encoding(DictEncoding.id()) {
+                    Self::Writer(args.child_strategy.new_writer(&args.ctx, &args.dtype)?)
+                } else if args.options.repeat {
```

Comment:
I think you should _always_ repeat now that you have the check above to see if dict encoding is applicable